feat: removed pkg_resources from all test files and moved importlib into pandas extra

kiraksi opened the pull request:
In this PR
- removed all instances of pkg_resources in tests and introduced importlib instead
- similar to https://github.com/googleapis/python-bigquery-storage/pull/706
- should resolve any further issues we had stemming from pkg_resources

Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [X] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/python-bigquery/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [X] Ensure the tests and linter pass
- [X] Code coverage does not decrease (if any source code was changed)
- [X] Appropriate docs were updated (if necessary)

Fixes #1725 🦕
Fixes #313 🦕

parthea committed:
feat: Introduce compatibility with native namespace packages

parthea committed:
Update copyright year

kiraksi committed:
removed pkg_resources from all test files and moved importlib into pandas extra

kiraksi committed:
feat: removed pkg_resources from all test files and moved importlib into pandas extra

kiraksi committed:
Adding no cover tag to test code

kiraksi committed:
reformatted with black

parthea commented on tests/system/test_pandas.py:
```diff
@@ -41,12 +45,12 @@
+    PANDAS_INSTALLED_VERSION = "0.0.0"
 
-PANDAS_INT64_VERSION = pkg_resources.parse_version("1.0.0")
+PANDAS_INT64_VERSION = "1.0.0"
```
```suggestion
```


Instead of 
```
@pytest.mark.skipif(
    PANDAS_INSTALLED_VERSION < PANDAS_INT64_VERSION,
    reason="Only `pandas version >=1.0.0` is supported",
)
```
we could have
```
@pytest.mark.skipif(
    PANDAS_INSTALLED_VERSION[0:2] in ["0."],
    reason="Only `pandas version >=1.0.0` is supported",
)
```

parthea commented on tests/unit/test__pandas_helpers.py:
```diff
@@ -57,13 +61,12 @@
-    PANDAS_INSTALLED_VERSION = pkg_resources.parse_version("0.0.0")
+if pandas is not None:  # pragma: NO COVER
+    PANDAS_INSTALLED_VERSION = metadata.version("pandas")
+else:  # pragma: NO COVER
```
Previously we didn't have `# pragma: NO COVER`. Is it needed?

kiraksi replied:
When I ran the kokoro build test it was giving me a failure on coverage in a test file (not this one), so I thought to add the tag everywhere just in case. It may have just been an issue with that test file, so I can remove it here and other places if needed!

kiraksi committed:
undo revert

parthea reviewed (approved):
LGTM. Added one observation for potentially unnecessary version checks in test files.

tswast committed:
perf: use the first page a results when `query(api_method="QUERY")` (#1723)

* perf: use the first page a results when `query(api_method="QUERY")`

* add tests

* respect max_results with cached page

* respect page_size, also avoid bqstorage if almost fully downloaded

* skip true test if bqstorage not installed

* coverage

tswast committed:
fix: ensure query job retry has longer deadline than API request deadline (#1734)

In cases where we can't disambiguate API failure from job failure,
this ensures we can still retry the job at least once.

chalmerlowe commented on tests/system/test_pandas.py:
```diff
@@ -311,7 +313,7 @@ def test_load_table_from_dataframe_w_automatic_schema(bigquery_client, dataset_i
 
 @pytest.mark.skipif(
-    PANDAS_INSTALLED_VERSION < PANDAS_INT64_VERSION,
+    PANDAS_INSTALLED_VERSION[0:2] in ["0."],
     reason="Only `pandas version >=1.0.0` is supported",
```
I tend to agree, we probably could simplify some of this. I am not sure that should hold up this PR.

kiraksi replied:
I can make this change today as well before I planned on merging

tswast committed:
fix: `load_table_from_dataframe` now assumes there may be local null values (#1735)

Even if the remote schema is REQUIRED

Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [ ] Make sure to open an issue as a [bug/issue](https://togithub.com/googleapis/python-bigquery/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [ ] Ensure the tests and linter pass
- [ ] Code coverage does not decrease (if any source code was changed)
- [ ] Appropriate docs were updated (if necessary)

Fixes #1692 🦕

kiraksi committed:
chore: standardize samples directory - delete unneeded dependencies (#1732)

* chore: standardize samples directory = delete unneeded dependencies

* Removed unused import for linter

kiraksi committed:
fix: move grpc, proto-plus and protobuf packages to extras (#1721)

* chore: move grpc, proto-plus and protobuff packages to extras

* formatted with black

* feat: add `job_timeout_ms` to job configuration classes (#1675)

* fix: adds new property and tests

* 🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

* updates docs to correct a sphinx failure

* Updates formatting

* Update tests/system/test_query.py

* 🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

* Update google/cloud/bigquery/job/base.py

* updates one test and uses int_or_none

* Update tests/system/test_query.py

testing something.

* Update tests/system/test_query.py

* testing coverage feature

* 🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

* minor edits

* tweaks to noxfile for testing purposes

* add new test to base as experiment

* adds a test, updates import statements

* add another test

* edit to tests

* formatting fixes

* update noxfile to correct debug code

* removes unneeded comments.

---------

Co-authored-by: Owl Bot <gcf-owl-bot[bot]@users.noreply.github.com>

---------

Co-authored-by: Chalmer Lowe <chalmerlowe@google.com>
Co-authored-by: Owl Bot <gcf-owl-bot[bot]@users.noreply.github.com>
Co-authored-by: Tim Swast <swast@google.com>

kiraksi committed:
remove unnecessary version checks

kiraksi committed:
undo bad commit, remove unneeded version checks

kiraksi committed:
Revert "undo bad commit, remove unneeded version checks"

This reverts commit 5c82dcf85f5fd73ad37ab7805b85b88a65351801.

kiraksi committed:
Revert "remove unnecessary version checks"

This reverts commit 9331a7e034de1c651c5ebc454f38f602d612a03d.

kiraksi committed:
revert bad changes, remove pkg_resources from file

kiraksi committed:
after clarification, reimplement changes and ignore 3.12 tests

kiraksi committed:
reformatted with black

kiraksi committed:
Merge branch 'main' into remove-pkgresources

kiraksi committed:
removed minimum check

chalmerlowe commented on tests/system/test_pandas.py:
```diff
@@ -1043,9 +1042,7 @@ def test_list_rows_max_results_w_bqstorage(bigquery_client):
-@pytest.mark.skipif(
-    PANDAS_INSTALLED_VERSION >= pkg_resources.parse_version("2.0.0"), reason=""
-)
+@pytest.mark.skipif(PANDAS_INSTALLED_VERSION[0:2] not in ["0.", "1."], reason="")
```
This has several elements in the list so it is a bit easier to grok what is happening here.

chalmerlowe reviewed (approved):
LGTM

parthea commented on tests/system/test_pandas.py:
```diff
@@ -342,7 +341,7 @@ def test_load_table_from_dataframe_w_nullable_int64_datatype(
 
 @pytest.mark.skipif(
-    PANDAS_INSTALLED_VERSION < PANDAS_INT64_VERSION,
+    PANDAS_INSTALLED_VERSION[0:2].startswith("0.")
     reason="Only `pandas version >=1.0.0` is supported",
```
```suggestion
    PANDAS_INSTALLED_VERSION[0:2].startswith("0."),
    reason="Only `pandas version >=1.0.0` is supported",
```

kiraksi committed:
updated pandas installed version check

parthea commented:
Good catch @tswast ! I can produce the issue when using an editable install using egg as it appears in the [build log](https://fusion2.corp.google.com/invocations/2be1e8d5-99f1-41aa-97e4-4cabb94bd441/targets/bigframes%2Fpresubmit%2Fe2e/log) 

When I install using an editable install using egg, the import fails
```
pip install -e git+https://github.com/googleapis/python-bigquery.git@1f4ebb1eca4f9380a31172fc8cb2fae125f8c5a2#egg=google_cloud_bigquery --ignore-installed
```

When I install without an editable install, the import succeeds
```
pip install https://github.com/googleapis/python-bigquery/archive/main.zip --ignore-installed
```

I can produce the same issue with google auth. The import of `google.auth` fails when I use
```
pip install -e git+https://github.com/googleapis/google-auth-library-python.git@7ab0fced0008ad9283c955b7dd7e2b7db55ae5e7#egg=google_auth --ignore-installed`
```

but not with
```
pip install https://github.com/googleapis/google-auth-library-python/archive/main.zip
```

The change to move to native namespace packages was [released in July](https://github.com/googleapis/google-auth-library-python/releases/tag/v2.22.0) for `google-auth`. 

We should investigate whether we want to block on this issue before reverting the change as we do want to support python 3.12 without using `pkg_resources`.  

We could also consider the change to move to native namespace packages as a potential breaking change and bump the major version.

tswast commented:
Thanks for the investigation! I think we can remove the "-e" in the BigFrames tests.
